### PR TITLE
feat(next-core): relay transform plugin

### PR DESCRIPTION
**Pull request by kwonoj**
#### What
Part 1 for WEB-848. 

This PR implements initial path to enable relay compiler option in Turbopack. PR is based on approach WEB-955, that passing custom plugin transform itself instead of trying to pass down configuration into turbopack. In result, this PR requires counterpart turbopack PR at https://github.com/vercel/turbo/pull/4721.

PR also refactors next-shared's transform bit, as we have grown number of custom transforms.

Unfortunately there are some runtime errors with this transforms, so it is not possible to use relay actually yet. WEB-956 tracks this. (https://github.com/swc-project/plugins/pull/179)

### CODE REVIEW

File: packages/next-swc/crates/next-core/src/next_shared/transforms/relay.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            // https://github.com/vercel/next.js/blob/ea472e8058faea8ebdab2ef6d3aab257a1f0d11c/packages/next/src/build/webpack-config.ts#L792
+            None,
+        ));
+
+        let is_typescript = match &self.config.language {
+            RelayLanguageConfig::TypeScript => true,
+            _ => false,
+        };
+
+        program.visit_mut_with(&mut resolver(
```
Comment (kwonoj) [inline packages/next-swc/crates/next-core/src/next_shared/transforms/relay.rs:92]: @kdy1 It doesn't seem like this is the correct way.

I tried to pass `ctx.unresolved_mark` and let it be applied, like

```
let ctxt = SyntaxContext::empty().apply_mark(unresolved_mark);
let ident = Ident::new("require".into(), Span::new(...).with_ctxt(ctxt.unwrap()));
```

But weirdly it gives error

```
    Panic: PanicInfo { payload: Any { .. }, message: Some(assertion failed: `(left != right)`
      left: `Mark(0)`,
     right: `Mark(0)`), location: Location { file: "/Users/ojkwon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/swc_common-0.31.5/src/syntax_pos/hygiene.rs", line: 375, col: 13 }, can_unwind: true }
```

Probably I'm doing something wrong?

**Review by sokra — CHANGES_REQUESTED** [review]

File: packages/next-swc/crates/next-core/src/next_shared/transforms/mod.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            span: s.span,
+            body: s
+                .body
+                .iter()
+                .map(|stmt| ModuleItem::Stmt(stmt.clone()))
+                .collect(),
+            shebang: s.shebang.clone(),
+        }),
+    }
+}
```
[the remark this answers is not in the document]
Comment (kwonoj) [reply packages/next-swc/crates/next-core/src/next_shared/transforms/mod.rs:69]: This is preserving existing logic, not a new one for the relay. May need further refactoring to remove the place it uses this.

**Comment by sokra** [comment]
needs rebase

**Review by sokra — APPROVED** [review]

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 27f6f932e7e18d7384de4c4c2d90329cc0ae045a

`pnpm testheadless test/e2e/middleware-dynamic-basepath-matcher/test/index.test.ts` 
- Middleware custom matchers basePath > should match query path

<details>
<summary>Expand output</summary>

  ● Middleware custom matchers basePath › should match query path

    expect(received).toBe(expected) // Object.is equality

    Expected: "another-page"
    Received: "random"

      51 |     await browser.elementById('linkelement').click()
      52 |     const anotherPagePath = await browser.elementById('router-path').text()
    > 53 |     expect(anotherPagePath).toBe('another-page')
         |                             ^
      54 |   })
      55 | })
      56 |

      at Object.<anonymous> (e2e/middleware-dynamic-basepath-matcher/test/index.test.ts:53:29)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by ijjk — APPROVED** [review]

**Pull request merged**